### ISSUE TITLE
Inconsistent treatment of millisecond-precision `time` for instant queries

### ISSUE THREAD

**Issue by dandv** [VictoriaMetrics/VictoriaMetrics#5796]
### Describe the bug

The documentation for instant queries [states](https://docs.victoriametrics.com/keyconcepts/#instant-query) that the `timestamp` [is parsed? should be given? - not clear] "in second precision". But the hyperlinked page for [timestamp](https://docs.victoriametrics.com/single-server-victoriametrics/#timestamp-formats) mentions milliseconds.

It's unclear how the `time` parameter is parsed to "second precision" when a millisecond-precision time is passed to instant queries.

**BUG 1:** this should be clarified in the docs.

**BUG 2:** This has implications in rollup functions, when points can be included or not when calculating durations as (end - start + 1ms).

For example if one wants to `count_over_time` the points between a `start` and an `end`, an algorithm would be:

    duration = end.getTime() - start.getTime() + 1;
    query = 'count_over_time(' + seriesSelector + '[' + duration + '])';

The problem is that this algorithm will break in certain situations, and it's unclear when (see below).

### To Reproduce

In the steps below, only the last 3 digits of the `time` parameter will be changed.

1. Create a series with two points 100ms apart:

       curl http://localhost:8428/api/v1/import/prometheus -d 'series1{label="foo"} 10 1707123456700
       series1{label="foo"} 20 1707123456800'

2. [Wait a few seconds](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5555), then verify that both points were created correctly:

        export VMQURL=http://localhost:8428/api/v1/query
        curl $VMQURL -d 'query=series1[101ms]' -d 'time=1707123456800' | jq  # values 10 and 20

3. Fetch the last point at its timestamp with step 1ms:

       curl -s $VMQURL -d 'query=series1' -d 'time=1707123456800' -d 'step=1ms' | grep 20  # found

4. Set the `time` to 199ms past it and keep step 1ms. The point is still found, presumably because of the ["second precision" of the `time` parameter for instant queries](https://docs.victoriametrics.com/keyconcepts/#instant-query):

       curl -s $VMQURL -d 'query=series1' -d 'time=1707123456999' -d 'step=1ms' | grep 20  # found

5. Now repeat with the first point at its own timestamp:

       curl -s $VMQURL -d 'query=series1' -d 'time=1707123456700' -d 'step=1ms' | grep 10  # found

6. Now let's set `time` to only 5ms past the timestamp of the first point, since even 199ms worked for the second point. Surprise, the point isn't returned :boom::

       curl -s $VMQURL -d 'query=series1' -d 'time=1707123456705' -d 'step=1ms' | grep 10  # nothing!

7. But, 4ms works: 🤨🤔

       curl -s $VMQURL -d 'query=series1' -d 'time=1707123456704' -d 'step=1ms' | grep 10  # found

8. Now let's fetch both points by passing a duration into the instant query: the difference between the two timestamps, plus 1ms (to include both points), so 101ms:

       curl -s $VMQURL -d 'query=series1[101ms]' -d 'time=1707123456800' -d 'step=1ms' | jq  # both found, correct

But, with different dates, the (end - start + 1ms) formula may break. Let's create a new series:

    curl http://localhost:8428/api/v1/import/prometheus -d 'series2{label="foo"} 10 1638564958042
    series2{label="foo"} 20 1642801537346'

The difference in ms between the two timestamps is 4236579304. However, an instant query with a duration of that difference + 1ms, anchored at the time of the last point, will only produce the last point:

    curl -s $VMQURL -d 'query=series2[4236579305ms]' -d 'time=1642801537346' | jq  # wrong, "10" not found

`count_over_time` will return `1` instead of `2`:

    curl -s $VMQURL -d 'query=count_over_time(series2[4236579305ms])' -d 'time=1642801537346' | jq  # wrong: 1

Adding 1ms to the duration (`4236579306ms`) does return 2. Is this some precision error? Is it necessary in other situations to add 3? Or 999? What if that includes unwanted points? For certain types of data/use cases, this may not be acceptable.

### Version

```
     First noticed in: victoria-metrics-20240130-205929-tags-v1.97.0-0-gfcc8b14f8
         Confirmed in: victoria-metrics-20240425-145433-tags-v1.101.0-0-g5334f0c2c
Still reproducible in: victoria-metrics-20241021-132655-tags-v1.105.0-0-g05f6ea621d
```

### Used command-line flags

-retentionPeriod=100y -search.latencyOffset=1s

**Comment by dandv** [comment]
I still see this issue in 1.105.0. I've updated the bug description to hopefully make it clearer.

Any chance you could take another look @Amper ?

**Comment by hagen1778** [comment]
@rtm0 could you please take a look?

**Comment by rtm0** [comment]
Hi @dandv, thanks for reporting these incosistencies. Some of them are expected behavior and we will need to update our documentation to reflect them. Others are indeed bugs and I've been working to fix them. Please find the details below.

> 2. Wait a few seconds, then verify that both points were created correctly:

If you don't want to wait, you can force-flush data (for experimenting only, do not use in prod):

```
curl http://localhost:8428/internal/force_flush
``` 

> 4. Set the time to 199ms past it and keep step 1ms. The point is still found, presumably because of the "second precision" of the time parameter for instant queries:

This is not because of second-precision, VictoriaMetrics can accept time in milliseconds, the documentation needs to be fixed. The result includes the last sample because internally VictoriaMetrics adds `5m` before the start time. That is, the actual start time is not `time - 1ms` but `time - 1ms - 5m`.

This is due to how `range queries` work. A range query corresponds to a set of instant queries. Which are executed at `start+step`, `start+2*step`, ..., `start+N*step`; where `N = floor[(end-start)/step]`.  If we want the result to be drawn (by VMUI or Grafana) as a line, the result must include a sample for each instant query within `[start..end]`. Points in time that miss a sample will be shown on the gragh as a gap. And the extreme case of it is a collection of dots that are not connected.

Common sense suggests that the result of an instant query is searched within the `[start-step..step]` interval. But there can be cases when the step is too short. For example, becase it is shorter than `scrape interval` or because of the data simply wasn't scraped due to the scraper being too busy or a network issue. Therefore the search interval for an instant query is unconditionally extended by 5 minutes, resulting in `[start-step-5m .. step]`.  

Now, that could make sense for range queries but how does all this apply to instant queries? After all, an instant query result is a single sample and gaps do not apply to it. It would be correct to expect no results for your query. However, we've decided to to keep the above behavior in intant queries as well for consistency. So the result you see is the expected one - VM unconditionally looks back for a data point on 5m + step range. A workaround would be setting `-search.minStalenessInterval=1ms`  or using rollup funcitons with explicit look-behind window like `last_over_time(series1[1ms])`. I will update docs.

Why not just substract 5m from time without also substracting step? Because sometimes step can be greater than 5m. Current implementation is to substract both, but we may change it to `max(step, 5m)` in future.


 > 6. Now let's set time to only 5ms past the timestamp of the first point, since even 199ms worked for the second point. Surprise, the point isn't returned 💥:
>  7. But, 4ms works: 🤨🤔

It is a bug. Based on the explanations above, the sample must be returned in both cases. This is possibly because VictoriaMetrics also performs automatic adjustment of the lookbehind window. The lookbehind window of `start-step-5m` will be used for the first 20 instant queries from the [start..end] range. For the subsecuent instant queries, the window size will be `calculated as the median of the intervals between the first 20 raw samples in the requested time range` (see [Range Query docs](https://docs.victoriametrics.com/keyconcepts/#range-query)).

After discussing this intervally we've decided to remove this `20 sample madian` functionality to see if it simplifies things and fixes this bug. 

> The difference in ms between the two timestamps is 4236579304. However, an instant query with a duration of that difference + 1ms, anchored at the time of the last point, will only produce the last point:

This is also a bug and it has been fixed. See PRs linked to this issue.

**Comment by f41gh7** [comment]
The first part of the bugfix was included into [v1.108.0](https://docs.victoriametrics.com/changelog/#v11080), [v1.102.9](https://docs.victoriametrics.com/changelog/#v11029)  and [v1.97.14](https://docs.victoriametrics.com/changelog/#v19714)

**Comment by hagen1778** [comment]
The [second part of the bugfix](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7767) was merged and will be included into the next release.

**Comment by f41gh7** [comment]
The bugfix was included into [v1.108.1](https://docs.victoriametrics.com/changelog/#v11081)

It will be also back-ported into the next LTS releases

**Comment by zekker6** [comment]
The bugfix was included into [v1.102.10](https://docs.victoriametrics.com/changelog/#v110210) and [v1.97.15](https://docs.victoriametrics.com/changelog/#v19715).
